### PR TITLE
Refactoring for contracts v1

### PR DESCRIPTION
**Pull request by oskin1**

### CODE REVIEW

File: src/Cardano/Utils.hs [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+import           Ledger.Value
+import           Ledger.Ada
+import qualified PlutusTx.AssocMap as Map
+
+lovelaceSubtract :: Value -> Ada -> Value
+lovelaceSubtract vl Lovelace{getLovelace=v} =
+    vl <> negValue
+  where
+    negValue = Value $ Map.fromList [(adaSymbol, Map.fromList [(adaToken, negate v)])]
```
Comment (GusevTimofey) [inline src/Cardano/Utils.hs:11]: Probably use https://github.com/input-output-hk/plutus/blob/master/plutus-ledger-api/src/Plutus/V1/Ledger/Ada.hs#L86

File: src/ErgoDex/Amm/PoolActions.hs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+mkPoolActions :: PubKeyHash -> PoolActions
+mkPoolActions executorPkh = PoolActions
+  { runSwap    = runSwap' executorPkh
+  , runDeposit = runDeposit' executorPkh
+  , runRedeem  = runRedeem' executorPkh
+  }
+
+runSwap' :: PubKeyHash -> Confirmed Swap -> Confirmed Pool -> Either OrderExecErr (TxCandidate, Predicted Pool)
+runSwap' executorPkh (Confirmed swapOut Swap{swapExFee=ExFeePerToken{..}, ..}) (Confirmed poolOut pool) = do
```
Comment (Bromel777) [inline src/ErgoDex/Amm/PoolActions.hs:40]: maybe also add check for compliance of the output with the pool parameters?

File: src/ErgoDex/Amm/PoolActions.hs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        initValue   = fullTxOutValue swapOut
+        rewardValue = (assetAmountValue quoteOutput) <> (lovelaceSubtract initValue (Ada.Lovelace executorFee))
+
+    outputs = [nextPoolOut, rewardOut, executorOut]
+
+  when (getAmount quoteOutput < swapMinQuoteOut) (Left PriceTooHigh)
+
+  Right (TxCandidate inputs outputs, pp)
+
+runDeposit' :: PubKeyHash -> Confirmed Deposit -> Confirmed Pool -> Either OrderExecErr (TxCandidate, Predicted Pool)
```
Comment (Bromel777) [inline src/ErgoDex/Amm/PoolActions.hs:73]: why not Plutus.V1.Ledger.Address?

File: src/ErgoDex/Amm/Orders.hs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  , depositPair      :: (AssetEntry, AssetEntry)
+  , depositExFee     :: ExFee
+  , depositRewardPkh :: PubKeyHash
+  } deriving (Show, Eq)
+
+instance FromLedger Deposit where
+  parseFromLedger fout@FullTxOut{fullTxOutDatum=(Just (Datum d)), ..} =
+    case fromBuiltinData d of
+      (Just DepositDatum{..}) ->
+        case excludeAdaFlattened fullTxOutValue of
```
Comment (Bromel777) [inline src/ErgoDex/Amm/Orders.hs:60]: Is the logic will be correct if one of the assets is ada?

File: src/ErgoDex/Amm/PoolActions.hs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  let
+    poolIn  = FullTxIn poolOut Pay2Script (Just $ Redeemer $ toBuiltinData P.Deposit)
+    orderIn = FullTxIn depositOut Pay2Script (Just unitRedeemer)
+    inputs  = [poolIn, orderIn]
+
+    (inX, inY) =
+        bimap entryAmount entryAmount $
+          if (assetEntryClass $ fst depositPair) == (unCoin poolCoinX)
+          then depositPair
+          else (snd depositPair, fst depositPair)
```
Comment (Bromel777) [inline src/ErgoDex/Amm/PoolActions.hs:84]: you can use swap from here https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Tuple.html#v%3aswap

File: src/ErgoDex/Types.hs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+assetAmountValue :: AssetAmount a -> Value
+assetAmountValue AssetAmount{getAsset=Coin ac, getAmount=Amount v} = assetClassValue ac v
+
+assetAmountOf :: AssetEntry -> AssetAmount a
+assetAmountOf (AssetEntry (ac, v)) = AssetAmount (Coin ac) (Amount v)
+
+assetAmountCoinOf :: Coin a -> Integer -> AssetAmount a
+assetAmountCoinOf c v = AssetAmount c (Amount v)
+
+assetAmountPairOf :: (AssetEntry, AssetEntry) -> Coin a -> AssetAmount a
```
Comment (Bromel777) [inline src/ErgoDex/Types.hs:57]: redundant func

File: src/ErgoDex/Types.hs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import Prelude (Show, Eq, Integer, ($), (==), (<>), negate)
+
+import           Ledger
+import           Ledger.Value      (AssetClass(..), assetClassValueOf, assetClassValue, Value(..))
+import           PlutusTx.Numeric  (AdditiveSemigroup(..), MultiplicativeSemigroup(..))
+import qualified PlutusTx.AssocMap as Map
+
+import ErgoDex.Contracts.Types
+
+data Lovelace = Lovelace
```
Comment (Bromel777) [inline src/ErgoDex/Types.hs:12]: you can use https://github.com/input-output-hk/plutus/blob/edd3ce253a9080cd92b785f5a2b48f7d5b09e350/plutus-ledger-api/src/Plutus/V1/Ledger/Ada.hs#L58

File: src/ErgoDex/Amm/Pool.hs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+import ErgoDex.State
+import ErgoDex.Contracts.Types
+import ErgoDex.Contracts.Pool
+import ErgoDex.OffChain
+
+newtype PoolId = PoolId { unPoolId :: Coin Nft }
+  deriving (Show, Eq, Generic, FromJSON, ToJSON)
+
+data PoolFee = PoolFee
+  { poolFeeNum :: Integer
```
Comment (Bromel777) [inline src/ErgoDex/Amm/Pool.hs:23]: maybe use Amount here?

File: src/Cardano/Address.hs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+import qualified Ledger.Typed.Scripts as Scripts
+import qualified Plutus.V1.Ledger.Scripts as PlutusScripts
+import qualified Ledger.Scripts as LedgerScripts
+import qualified Cardano.Api as Script
+import Data.Text
+
+mkScriptAddress :: forall a . Scripts.TypedValidator a -> Text
+mkScriptAddress validatorInstance =
+    (Script.serialiseAddress . Script.makeShelleyAddress Script.Mainnet paymentCredential) Script.NoStakeAddress
```
Comment (Bromel777) [inline src/Cardano/Address.hs:13]: is it necessary for hardcoding Mainnet?

File: src/ErgoDex/Amm/Pool.hs (hunk 8 above)
Comment (oskin1) [reply src/ErgoDex/Amm/Pool.hs:24]: Its a numerator not an Amount.

File: src/ErgoDex/Amm/PoolActions.hs (hunk 3 above)
Comment (Bromel777) [reply src/ErgoDex/Amm/PoolActions.hs:77]: Cause inside you are using it only for address creation

File: src/Cardano/Address.hs (hunk 9 above)
Comment (oskin1) [reply src/Cardano/Address.hs]: Fixed.

**Review by GusevTimofey — APPROVED** [review]

**Pull request merged**